jcli/src/jcli_app/utils/open_api_verifier.rs

```diff
@@ -130,46 +171,70 @@ fn verify_path(
     req: &Request,
     body: &RestApiRequestBody,
 ) -> Result<(), PathError> {
-    let path = find_path_item(openapi, req.url().path())?;
-    verify_method(path, req, body).map_err(|source| PathError::MethodVerificationFailed {
-        source,
-        method: req.method().to_string(),
+    let (wildcards, path_item) = find_path_item(&openapi.paths, req.url().path())?;
+    verify_method(wildcards, path_item, req, body).map_err(|source| {
+        PathError::MethodVerificationFailed {
+            source,
+            method: req.method().to_string(),
+        }
     })
 }
 
-fn find_path_item<'a>(openapi: &'a OpenAPI, url: &str) -> Result<&'a PathItem, PathError> {
-    openapi
-        .paths
+fn find_path_item<'a>(
+    paths: &'a Paths,
+    url: &'a str,
+) -> Result<(PathWildcardValues<'a>, &'a PathItem), PathError> {
+    let (wildcards, path_item_ref) = paths
         .iter()
-        .find(|(path, _)| url_matches_path(url, &path))
-        .ok_or_else(|| PathError::NotFound)
-        .and_then(|(_, item)| unpack_reference_or(item).map_err(Into::into))
+        .find_map(|(path, path_item_ref)| {
+            url_matches_path(url, &path).map(|wildcards| (wildcards, path_item_ref))
+        })
+        .ok_or_else(|| PathError::NotFound)?;
+    let path_item = unpack_reference_or(path_item_ref)?;
+    Ok((wildcards, path_item))
 }
 
-fn url_matches_path(url: &str, path: &str) -> bool {
-    let url_segments = url.trim_matches('/').rsplit('/');
-    let mut path_segments = path.trim_matches('/').rsplit('/');
-    let segments_eq = url_segments
-        .zip(path_segments.by_ref())
-        .all(url_segment_matches_path_segment);
-    let path_exhausted = path_segments.next().is_none();
-    segments_eq && path_exhausted
+fn url_matches_path<'a>(url: &'a str, path: &'a str) -> Option<PathWildcardValues<'a>> {
+    let mut url_segments = url.trim_matches('/').rsplit('/');
+    let path_segments = path.trim_matches('/').rsplit('/');
+    let mut wildcards = HashMap::new();
+    for path_segment in path_segments {
+        let url_segment = url_segments.next()?;
+        match url_segment_matches_path_segment(url_segment, path_segment) {
+            SegmentMatch::None => return None,
+            SegmentMatch::Exact => continue,
+            SegmentMatch::Wildcard(wildcard) => wildcards.insert(wildcard, url_segment),
```

Comment:
Why is a hash table used here? What happens when there is a wildcard segment identical to another segment used elsewhere in the path?

Reply:
The key is wildcard name and wildcard names must be unique. The HashMap gives convenient access to wildcard values taken from URL using wildcard names taken from OpenAPI definition.